gnc/sub8_perception/sub8_vision_tools/labelling/manual_grabcut.py

```diff
@@ -75,67 +80,124 @@ def wait_for_key(self, keys):
+        print 'Keys:'
+        print '\tPress "space" to skip the current segmentation'
+        print '\tPress "q" to quit and save all segmentations including this'
+        print '\tPress w record this segmentation and move on to the next image'
         key = self.wait_for_key('qnzw ')
         if key in [' ', 'q']:
-                return None, key
+                return None, None, key
 
-        display_int = (255 * display) / np.max(display).astype(np.uint8)
-        return out_mask, key
+        return out_mask, return_mask, key
 
     def save(self, mask):
         data = {"image": self.image, 'classes': mask}
         pickle.dump(data, open("save.p", "wb"))
 
 
 if __name__ == '__main__':
-    # img = cv2.imread('/home/jacob/Downloads/buoy.jpg')[::10, ::10, :]
-    bag = '/home/jacob/catkin_ws/src/Sub8/gnc/sub8_perception/data/bag_test.bag'
+    usage_msg = ("Pass the path to a bag, and we'll crawl through the images in it")
+    desc_msg = "A tool for making manual segmentation fun!"
+
+    parser = argparse.ArgumentParser(usage=usage_msg, description=desc_msg)
+    parser.add_argument(dest='bag',
+                        help="The topic name you'd like to listen to")
+    parser.add_argument('--topic', type=str, help="Name of the topic to use")
+    parser.add_argument('--append', type=str, help="Path to a file to append to")
+    parser.add_argument('--output', type=str, help="Path to a file to output to (and overwrite)",
+                        default='segments.p')
+    parser.add_argument('--brush_size', type=int, help="Brush size",
+                        default=3)
+
+    args = parser.parse_args(sys.argv[1:])
+
+    bag = args.bag
     bc = bag_crawler.BagCrawler(bag)
 
-    data = []
+    if args.append is None:
+        print 'Creating new pickle'
+        data = []
+    else:
+        data = pickle.load(open(args.append, "rb"))
 
+    print bc.image_topics[0]
+    print bc.image_topics[0]
     last_mask = None
-    for image in bc.crawl(topic=bc.image_topics[0]):
-        p = Picker(image, brush_size=3, initial_mask=last_mask)
-        last_mask, key = p.segment()
+    num_imgs = len(data)
+
+    if args.topic is not None:
+        assert args.topic in bc.image_topics, "{} not in the bag".format(args.topic)
+        print 'Crawling topic {}'.format(args.topic)
+        crawl = bc.crawl(topic=args.topic)
+    else:
+        crawl = bc.crawl(topic=bc.image_topics[0])
+
+    for image in crawl:
+        num_imgs += 1
+        print 'On image #{}'.format(num_imgs)
+        p = Picker(image, brush_size=args.brush_size, initial_mask=last_mask)
+        last_mask, last_targets, key = p.segment()
         if key == 'q':
             break
         elif key == ' ':
             print 'ignoring image'
             continue
 
         print 'recording'
-        data.append((image, last_mask))
-
-    pickle.dump(data, open('segments.p', 'wb'))
+        data.append((image, last_targets))
```

Comment:
I'm assuming this gets run on big datasets, it'd be a shame if something were to happen and crash the program midway into the job. Can you save temp pickles along the way or does writing a pickle file take too long?